org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServers.java

```diff
@@ -227,34 +228,48 @@ public static class LanguageServerDocumentExecutor extends LanguageServers<Langu
 			return VersionedEdits.toVersionedEdits(this, edits);
 		}
 
+		private @Nullable CompletableFuture<LanguageServer> connect (@Nullable LanguageServerWrapper w){
+			if (w != null) {
+				try {
+					return w.connect(document);
+				} catch (IOException e) {
+					LanguageServerPlugin.logError(e);
+				}
+			}
+			return null;
+		}
 		/**
-		 * Test whether this server supports the requested <code>ServerCapabilities</code>, and ensure
-		 * that it is connected to the document if so.
-		 *
-		 * NB result is a future on this <emph>wrapper</emph> rather than the wrapped language server directly,
-		 * to support accessing the server on the single-threaded dispatch queue.
+		 * Returns a {@link CompletableFuture} that assures that the document is connected to the server before completing.
 		 */
-		@NonNull CompletableFuture<@Nullable LanguageServerWrapper> connectIf(@NonNull LanguageServerWrapper wrapper) {
-			return wrapper.getInitializedServer().thenCompose(server -> {
-				if (server != null && getFilter().test(wrapper.getServerCapabilities())) {
-					try {
-						return wrapper.connect(document);
-					} catch (IOException ex) {
-						LanguageServerPlugin.logError(ex);
-					}
-				}
-				return CompletableFuture.completedFuture(null);
-			}).thenApply(server -> server == null ? null : wrapper);
+		private @NonNull CompletableFuture<@Nullable LanguageServerWrapper> connectIf(@NonNull CompletableFuture<@Nullable LanguageServerWrapper> wrapper) {
+			return wrapper.thenCombine(wrapper.thenApply(this::connect), (w, server) -> w);
```

Comment:
this thenCombine looks cumbersome to me, but I did not find anything better. @ahmedneilhussain , is there anything simpler to achieve this? I am quite struggling with the CompletableFuture API, so probably there is something simple which I have missed